Add RTL Support

zakariaf opened the pull request:
# Add RTL Support to Jekyll Theme Chirpy

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Improvement (refactoring and improving code)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation update

## Description

This PR adds comprehensive Right-to-Left (RTL) language support to the Chirpy theme. It enables the theme to automatically detect RTL languages and apply appropriate styling, making it accessible to users of Arabic, Kurdish, Persian, Urdu, and other RTL languages.

The implementation:
- Automatically detects RTL languages based on the page's `lang` attribute
- Applies appropriate direction and styling for RTL content
- Provides language-specific fonts for various RTL languages
- Maintains all existing LTR functionality without changes
- Includes configuration options to customize RTL behavior

### Key components:

1. **Language/Direction Detection**: Added to `lang.html` to detect RTL languages
2. **RTL Stylesheet**: Created `rtl.scss` with comprehensive RTL styling
3. **RTL Font Support**: Added `rtl-fonts.scss` for language-specific fonts
4. **Example Implementation**: Provided Persian locale file and sample content

### New Files
- `_sass/rtl.scss`: Contains RTL-specific styles for layout and UI elements
- `_sass/rtl-fonts.scss`: Defines and applies language-specific fonts for RTL languages
- `_includes/rtl-font-loader.html`: HTML-based font loader for faster RTL font loading
- `_data/locales/*`: Locale files as reference implementation
    - `fa-IR.yml`: Persian (Modified)
    - `ku-IQ.yml`: Kurdish(Sorani)
    - `ur-PK.yml`: Urdu
    - `ps-AF.yml`: Pashto
    - `dv-MV.yml`: Dhivehi
- `docs/RTL-SUPPORT.md`: Documentation for RTL support

### Modified Files
- `_sass/main.scss`: Added import for rtl.scss
- `_includes/lang.html`: Added RTL detection logic
- `_layouts/default.html`: Added dir attribute to HTML tag
- `_config.yml`: Added RTL configuration options

This change enhances accessibility for millions of users worldwide who read and write in RTL languages without affecting any existing functionality.

## Additional context
This implements a non-breaking enhancement that makes the theme more accessible to global users.

Note: The implementation requires a small addition to `head.html` to include the RTL font loader, which I've documented in the PR.

zakariaf commented:
@cotes2020 This closes the issue https://github.com/cotes2020/jekyll-theme-chirpy/issues/1928 too

zakariaf commented:
I'm running my Persian blog based on these changes, and it seems works perfectly. https://zakariafatahi.ir/

zakariaf commented:
hello @cotes2020  I just wanted to check if there’s anything else needed from my side to move this PR forward or if there are any concerns about merging it.

cotes2020 commented:
>... I just wanted to check if there’s anything else needed from my side to move this PR forward or if there are any concerns about merging it.

Hi there, this is an excellent PR and thank you for your effort! Since there are quite a lot of changes, I’ll need to find some time to review it. Please don’t worry.

cotes2020 reviewed (changes requested):
Sorry for leaving this PR pending for so long. Here is my initial review:

1. Defining different webfonts for specific languages is too complex and will make the project harder to maintain. Therefore, for languages other than English, it's better to use the system's default fonts.

2. If you want to achieve automatic enabling of the RTL (right-to-left) attribute by intelligently identifying the language, then the configuration file does not need new settings. The Liquid code can identify which languages are RTL.

zakariaf commented:
Thank you for the review!

Regarding the font: I agree it's better to keep things simple for maintainability. That said, the default English font renders quite poorly for RTL languages in practice. For now, I’ve removed the custom fonts as suggested, but I’ll experiment with a simpler approach in a future PR—likely by adding something like this to config.yml file:

```yaml
# Font Configuration
fonts:
  # Base font for body text
  base: ["Source Sans Pro", "Microsoft Yahei", sans-serif]
  # Font for headings
  heading: ["Lato", "Microsoft Yahei", sans-serif]
  # Optional: Google Fonts to load
  google_fonts:
    # - "Roboto:400,700"
    # - "Montserrat:700"
```

I'll open a separate PR for this to keep things clean.

As for RTL detection via Liquid: I had tested that path, but couldn’t find a reliable way to automatically detect RTL languages purely within Liquid. I’ve simplified the implementation significantly in this PR to make it easier to manage.

Thanks again!

Nicolai9852 commented:
Hey @zakariaf 

I just saw your RTL - it looks really awesome! I did however notice one thing - not sure if it's intentional or not

In the default LTR version, the button to copy the header link is right next to the title
In this RLT it is by the far end to the left side.

From LRT:
![billede](https://github.com/user-attachments/assets/73d5a839-fe61-4fc2-ac9e-fa3c9c21a5f5)

RLT:
![billede](https://github.com/user-attachments/assets/5332ab55-d042-41bb-85f3-f56c5e18c55a)

cotes2020 reviewed (changes requested):
> ... but couldn’t find a reliable way to automatically detect RTL languages purely within Liquid. 

Use this code to automatically detect RTL:

```liquid
{% assign _lang_code = lang | split: '-' | first %}

{%- case _lang_code -%}
  {%- when 'ar', 'fa', 'he', 'iw', 'kd', 'pk', 'ps', 'ug', 'ur', 'yi' -%}
    {% assign dir = 'rtl' %}
{%- endcase -%}
```

Additionally, please move the newly added files in `_data/locales` to a separate PR. The file `_posts/2025-04-20-rtl-test-persian.md` doesn't need to be included in the repository either.

zakariaf commented:
Hey @Nicolai9852, thanks for catching that! I think it was a mistake — I just fixed it. Appreciate you pointing it out!

zakariaf commented:
@cotes2020 Thanks so much for the suggestions! I’ve applied all of them. Also, I just created another PR for the new locale files:

[https://github.com/cotes2020/jekyll-theme-chirpy/pull/2415](https://github.com/cotes2020/jekyll-theme-chirpy/pull/2415)

cotes2020 commented on _sass/base/_rtl.scss:
```diff
@@ -0,0 +1,551 @@
+    padding-left: 0 !important;
+    padding-right: 0 !important;
+  }
+
+  .ps-2 {
+    padding-left: 0 !important;
+    padding-right: 0.5rem !important;
+  }
+
+  .ps-3 {
+    padding-left: 0 !important;
+    padding-right: 1rem !important;
+  }
+
+  .ps-4 {
+    padding-left: 0 !important;
+    padding-right: 1.5rem !important;
+  }
+
+  .ms-auto {
+    margin-left: 0 !important;
+    margin-right: auto !important;
+  }
+
+  .px-md-3 {
+    @include bp.md {
+      padding-right: 1rem !important;
+      padding-left: 1rem !important;
+    }
+  }
+
+  /* Text alignment in RTL */
+  .text-start {
+    text-align: right !important;
+  }
+
+  .text-end {
+    text-align: left !important;
+  }
+}
```
In addition, I’ve found several UI issues:

1. The dividing line between panel and main content should be positioned between the two.
2. On the _Archive_ page, the post list is not vertically aligned with the heading, and the vertical line of the "timeline" is also missing.

<img width="1512" alt="rtl" src="https://github.com/user-attachments/assets/fddcc741-ecb0-42f3-8222-66aedcc0c3a1" />

zakariaf replied:
I just fixed those, but here’s the result:

<img width="1229" alt="image" src="https://github.com/user-attachments/assets/3407ab05-0575-4175-970b-599c76e38221" />


This happens because I was using custom CSS classes and would have to modify more Bootstrap utilities for RTL.

However, after adding the official Bootstrap RTL stylesheet link to `head.html` and removing my custom classes, I got:

<img width="1229" alt="image" src="https://github.com/user-attachments/assets/96b69d5d-5d65-4400-9ca7-8b1094c670f2" />

In other words, after these changes I made, and by dropping the bootstrap custom classes and using Bootstrap’s RTL style, everything is fixed. My only question is: where’s the best place to include the RTL stylesheet? Should it go directly in `head.html`, or is there a better spot in our build or templating pipeline?

cotes2020 replied:
Take a look at the existing code, and you'll know where it was introduced. Just add a check for `dir=rtl` into the existing logic.

cotes2020 replied:
Another reminder: a large number of fragmented commits can make code review more difficult. In a PR, you can squash them into a single commit.

cotes2020 reviewed (changes requested):
The search bar and the trending tags and TOC in the panel are still not vertically aligned. Please test each page locally first. Thank you!

zakariaf commented:
I tried to check everything, I think all are fixed now.
